SYCL/ESIMD/api/simd_memory_access.cpp

```diff
@@ -0,0 +1,181 @@
+//==------- simd_memory_access.cpp  - DPC++ ESIMD on-device test -----------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+// REQUIRES: gpu
+// UNSUPPORTED: cuda
+// RUN: %clangxx -fsycl %s -o %t.out
+// RUN: %GPU_RUN_PLACEHOLDER %t.out
+//
+// The test checks functionality of the memory access APIs which are members of
+// the simd class.
+
+#include "esimd_test_utils.hpp"
+
+#include <CL/sycl.hpp>
+#include <CL/sycl/INTEL/esimd.hpp>
+#include <iostream>
+
+using namespace cl::sycl;
+
+template <typename T>
+using Acc =
+    accessor<T, 1, access_mode::read_write, access::target::global_buffer>;
+
+template <typename T, int N, bool IsAcc> struct Kernel;
+
+// Accessor-based kernel.
+template <typename T, int N> struct Kernel<T, N, true> {
+  Acc<T> acc;
+  Kernel(Acc<T> acc) : acc(acc) {}
+
+  void operator()(id<1> i) const SYCL_ESIMD_KERNEL {
+    using namespace sycl::INTEL::gpu;
+    const uint32_t ii = static_cast<uint32_t>(i.get(0));
+    simd<T, N> v;
+    const auto offset = ii * sizeof(v);
+    v.copy_from(acc, offset);
```

Comment:
We copy the source data into the original destination. If the `copy_from` and `copy_to` would be unimplemented the test will pass as well. Maybe we can test introduce a separate destination buffer and test copying `A -> B`?

Reply:
That is a good point!

Reply:
Actually, not quite. The test does:
```cpp
    v.copy_from(ptr + offset);
    v += simd<T, N>(ii * N, 1);
    v.copy_to(ptr + offset);
```
I.e. the written value depends both on the original value. So if either does not work test should fail. But the choice of '0' as the initial value is bad indeed, as there is a chance `v` can be initialized to zeroes even if `copy_from` does not work. I'll fix that.

Reply:
I mean, the point is still good :), as it lead to problem discovery, but it is fully valid.